graph/cycle.go

```diff
@@ -0,0 +1,119 @@
+// cycle.go
+// this file handle algorithm that related to cycle in graph
+// reference: https://en.wikipedia.org/wiki/Cycle_(graph_theory)
+// [kiarash hajian](https://github.com/kiarash8112)
+
+package graph
+
+import "strconv"
+
+func (g *Graph) HasCycle() bool {
+	//this implimetation refered as 3-color too
+	all := map[int]struct{}{}
+	visiting := map[int]struct{}{}
+	visited := map[int]struct{}{}
+
+	for v := range g.edges {
+		all[v] = struct{}{}
+	}
+
+	for current := range all {
+		if g.HasCycleHelper(current, all, visiting, visited) {
+			return true
+		}
+	}
+
+	return false
+
+}
+
+func (g Graph) HasCycleHelper(v int, all, visiting, visited map[int]struct{}) bool {
```

Comment:
```suggestion
func (g Graph) hasCycleHelper(v int, all, visiting, visited map[int]struct{}) bool {
```

This function is not needed by anything outside of this package and as such, should not be exported.